dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java

```diff
@@ -0,0 +1,180 @@
+import com.google.common.collect.Sets;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.bcel.Const;
+import org.apache.bcel.classfile.ClassParser;
+import org.apache.bcel.classfile.Constant;
+import org.apache.bcel.classfile.ConstantMethodref;
+import org.apache.bcel.classfile.ConstantNameAndType;
+import org.apache.bcel.classfile.ConstantPool;
+import org.apache.bcel.classfile.JavaClass;
+import org.apache.bcel.classfile.Method;
+
+/**
+ * This class reads Java class file to analyze following attributes:
+ *
+ * <ol>
+ *   <li>source (defined methods) via Method fields in the class, and</li>
+ *   <li>targets (what's attempted to be invoked) via ConstantPool field of the class</li>
+ * </ol>
+ */
+class ClassDumper {
+
+  private static List<ConstantPoolMethodref> listConstantPoolMethodRef(JavaClass javaClass) {
+    List<ConstantPoolMethodref> constantPoolMethodrefs = new ArrayList<>();
+    ConstantPool constantPool = javaClass.getConstantPool();
+    Constant[] constants = constantPool.getConstantPool();
+    List<Constant> methodrefConstants = Arrays.stream(constants)
+            .filter(Predicates.notNull())
+            .filter(constant -> constant.getTag() == Const.CONSTANT_Methodref)
+            .collect(Collectors.toList());
+    for (Constant constant : methodrefConstants) {
+      ConstantMethodref constantMethodref = (ConstantMethodref) constant;
+      String className = constantMethodref.getClass(constantPool);
+      int nameAndTypeIndex = constantMethodref.getNameAndTypeIndex();
+      Constant constantNameAndTypeRaw = constantPool.getConstant(nameAndTypeIndex);
+      if (!(constantNameAndTypeRaw instanceof ConstantNameAndType)) {
+        throw new RuntimeException(
+            "Failed to lookup nameAndType constant indexed " + nameAndTypeIndex);
+      }
+      ConstantNameAndType constantNameAndType = (ConstantNameAndType) constantNameAndTypeRaw;
+      String methodName = constantNameAndType.getName(constantPool);
+      String signature = constantNameAndType.getSignature(constantPool);
+      ConstantPoolMethodref methodref = new ConstantPoolMethodref(className, methodName, signature);
+      constantPoolMethodrefs.add(methodref);
+    }
+    return constantPoolMethodrefs;
+  }
+
+  /**
+   *  Lists all methodref entries defined in ConstantPool of class file
+   *
+   * @param classFileStream stream of a class file
+   * @param fileName name of the file that contains class
+   * @return List of all methodref entries defined in ConstantPool
+   * @throws IOException when there is problem in reading classFileStream
+   */
+  @VisibleForTesting
+  static List<ConstantPoolMethodref> listConstantPoolMethodref(InputStream classFileStream,
+      String fileName) throws IOException {
+    ClassParser parser = new ClassParser(classFileStream, fileName);
```

Comment:
I find it odd that we need to pass both the class file stream and the file name that the class file stream is derived from - do you know why `ClassParser` requires that?

Reply:
Ok, good to know.